Framework/Kernel/src/UsageService.cpp

```diff
@@ -106,15 +129,41 @@ void UsageServiceImpl::registerStartup() {
 
 /** registerFeatureUsage
  */
-void UsageServiceImpl::registerFeatureUsage(const std::string &type,
-                                            const std::string &name,
+void UsageServiceImpl::registerFeatureUsage(const FeatureType &type,
+                                            std::vector<std::string> name,
+                                            const bool internal) {
+  if (isEnabled()) {
+    std::lock_guard<std::mutex> _lock(m_mutex);
+
+    using boost::algorithm::trim;
+    std::for_each(std::begin(name), std::end(name),
+                  boost::bind(&trim<std::string>, _1, std::locale()));
+
+    using boost::algorithm::join;
+    std::string strName = join(name, SEPARATOR);
+
+    m_FeatureQueue.push(FeatureUsage(type, strName, internal));
+  }
+}
+
+void UsageServiceImpl::registerFeatureUsage(const FeatureType &type,
+                                            std::string name,
                                             const bool internal) {
   if (isEnabled()) {
     std::lock_guard<std::mutex> _lock(m_mutex);
+
+    boost::algorithm::trim(name);
     m_FeatureQueue.push(FeatureUsage(type, name, internal));
   }
 }
 
+void UsageServiceImpl::registerFeatureUsage(
+    const FeatureType &type, std::initializer_list<const char *> name,
+    const bool internal) {
+  registerFeatureUsage(type, std::vector<std::string>(name.begin(), name.end()),
```

Comment:
I wonder if we can avoid the vector construction at all? `std::initializer_list` can be iterated over but the problem is that each `const char *` cannot be trimmed in place. I'm wondering if we just drop the `trim`? It's pretty trivial in SQL to replace whitespace when running a query and that's really where the results will be viewed. If we remove the trimming we're just left with the join on both the vector version and the `initializer_list` version.

What do you think?

Reply:
The TrackedAction class uses the vector construction so seems worth keeping it. I think I can simplify the function you've commented on a bit by changing the std::initializer_list to use the type std::string rather than const char* though

Reply:
I will drop the trim as you suggest